2.0-PROPOSAL/README.md

```diff
@@ -0,0 +1,907 @@
+light of this, it is proposed that users be advised to designate (through
+labels) a _subset_ of their cluster's nodes for the execution of Brigade workers
+and jobs. Those nodes may optionally be tainted to repel non-Brigade workloads.
+
+### Securing Brigade
+
+In contrast to Brigade 1.x, the design proposed here does not have Brigade
+clients communicating directly with the Kubernetes API server. By requiring
+Brigade 2.0 clients to communicate with a new Brigade API instead, the implicit
+delegation of authentication and authorization to the Kubernetes API server
+comes to an end and must be replaced with other access control measures.
+
+#### Authentication
+
+Authentication is not, in and of itself onerous to implement, but tangential
+concerns such as user registration and password reset functionality can be. Such
+features would additionally increase the footprint of Brigade's infrastructure
+dependencies in that they rely on outbound email for purposes of identity
+verification. Wishing to avoid both the development effort and the additional
+dependencies, the author proposes utilizing [OpenID
+Connect](https://openid.net/connect/) in order to delegate authentication (of
+human users) to trusted identity providers such as Azure Active Directory or
+Google Identity Platform.
+
+The following sequence diagram depicts a human user logging into Brigade using
+the CLI. The initial request to log in returns a _unactivated_ bearer token as
+well as a URL that the human user can navigate to using their web browser in
+order to complete authentication using their trusted identity provider.
+Completion of that process results in bearer token _activation_. All subsequent
+user interactions with Brigade will present the bearer token as proof of
+authentication. Bearer tokens will expire periodically. (As a point of
+reference, the Azure CLI implements authentication in a similar fashion.)
+
+![Authentication Using OpenID Connect](authentication.png)
+
+N.B.: To prevent the dependency on an identity provider from becoming an
+impedance to evaluation (aka "tire kicking") and local development, as well as
+to facilitate initial setup of Brigade on remote clusters, it is proposed that
+"root" access to Brigade may be selectively enabled / disabled at the time of
+deployment / re-deployment / upgrade.
+
+N.B.: When any _new_ user authenticates with OpenID Connect, a new Brigade
+account is created automatically and federated with the user's identity from the
+trusted provider. Such a user initially has no roles assigned (see
+[authorization](#authorization) section). Roles may be granted by a user manager
+or project admin. It is proposed that users, once registered _cannot_ be deleted
+since nothing prevents a deleted user from re-registering. The benefits of
+_keeping_ but disabling the user account and thereby maintaining the relational
+integrity of the underlying data store seem to outweigh the finality of
```

Comment:
Is there going to be any problems with data retention laws here? People might want to be "fully removed"

Reply:
Admittedly, I don't know. I've been looking at "data retention" from the perspective of an enterprise that is required to _keep_ records and not from the opposite perspective where the right to disappear into the ether might be mandated by law. My understanding is that such laws are for consumer protection, however, and I don't see Brigade as a consumer-facing product.

It remains a good question. If we learn later that there is any kind of issue with this, I am pretty confident we can fix it.